File: android/src/main/kotlin/com/dkmads/dmp/Demographics.kt
package com.dkmads.dmp

import java.time.LocalDate
import java.time.Period
import java.time.format.DateTimeParseException

/** Canonical demographic buckets — mirrors `@dkmads/shared` / `packages/sdk-core`. */
object Demographics {
    val STANDARD_AGE_RANGES = listOf(
        "18-24", "25-34", "35-44", "45-54", "55-64", "65+", "unknown",
    )
    val STANDARD_GENDER_VALUES = listOf("male", "female", "other", "unknown")
    const val DEMOGRAPHIC_AGE_RANGE_KEY = "demographic.age_range"
    const val DEMOGRAPHIC_GENDER_KEY = "demographic.gender"
    const val MIN_AD_TARGET_AGE = 18

    fun ageFromDateOfBirth(dob: LocalDate, asOf: LocalDate = LocalDate.now()): Int {
        return Period.between(dob, asOf).years
    }

    fun ageRangeFromAge(age: Int): String {
        if (age < MIN_AD_TARGET_AGE) return "unknown"
        if (age <= 24) return "18-24"
        if (age <= 34) return "25-34"
        if (age <= 44) return "35-44"
        if (age <= 54) return "45-54"
        if (age <= 64) return "55-64"
        return "65+"
    }

    fun ageRangeFromDateOfBirth(dob: LocalDate, asOf: LocalDate = LocalDate.now()): String {
        return ageRangeFromAge(ageFromDateOfBirth(dob, asOf))
    }

    fun ageRangeFromDateOfBirth(dob: String, asOf: LocalDate = LocalDate.now()): String? {
        return try {
            ageRangeFromDateOfBirth(LocalDate.parse(dob), asOf)
        } catch (_: DateTimeParseException) {
            null
        }
    }

    fun normalizeAgeRange(value: String?): String? {
        val v = value?.trim() ?: return null
        return if (STANDARD_AGE_RANGES.contains(v)) v else null
    }

    fun normalizeGender(value: String?): String? {
        val v = value?.trim()?.lowercase() ?: return null
        return if (STANDARD_GENDER_VALUES.contains(v)) v else null
    }
}

File: ios/Sources/DKMadsDMP/Demographics.swift
import Foundation

/// Canonical demographic buckets — mirrors `@dkmads/shared` / `packages/sdk-core`.
public enum DMPDemographics {
    public static let standardAgeRanges = [
        "18-24", "25-34", "35-44", "45-54", "55-64", "65+", "unknown",
    ]
    public static let standardGenderValues = ["male", "female", "other", "unknown"]
    public static let demographicAgeRangeKey = "demographic.age_range"
    public static let demographicGenderKey = "demographic.gender"
    public static let minAdTargetAge = 18

    public static func ageFromDateOfBirth(_ dob: Date, asOf: Date = Date()) -> Int {
        let cal = Calendar.current
        var age = cal.component(.year, from: asOf) - cal.component(.year, from: dob)
        let monthDiff = cal.component(.month, from: asOf) - cal.component(.month, from: dob)
        if monthDiff < 0 || (monthDiff == 0 && cal.component(.day, from: asOf) < cal.component(.day, from: dob)) {
            age -= 1
        }
        return age
    }

    public static func ageRangeFromAge(_ age: Int) -> String {
        if age < minAdTargetAge { return "unknown" }
        if age <= 24 { return "18-24" }
        if age <= 34 { return "25-34" }
        if age <= 44 { return "35-44" }
        if age <= 54 { return "45-54" }
        if age <= 64 { return "55-64" }
        return "65+"
    }

    public static func ageRangeFromDateOfBirth(_ dob: Date, asOf: Date = Date()) -> String {
        let age = ageFromDateOfBirth(dob, asOf: asOf)
        if age < 0 { return "unknown" }
        return ageRangeFromAge(age)
    }

    public static func normalizeAgeRange(_ value: String) -> String? {
        let v = value.trimmingCharacters(in: .whitespacesAndNewlines)
        return standardAgeRanges.contains(v) ? v : nil
    }

    public static func normalizeGender(_ value: String) -> String? {
        let v = value.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        return standardGenderValues.contains(v) ? v : nil
    }
}
